src/matrixrtc/NewMembershipManager.ts

```diff
@@ -0,0 +1,945 @@
+        return this.joinConfig?.membershipExpiryTimeoutHeadroom ?? 5_000;
+    }
+    private computeNextExpiryActionTs(iteration: number): number {
+        return (
+            this.state.startTime +
+            this.membershipEventExpiryTimeout * iteration -
+            this.membershipEventExpiryTimeoutHeadroom
+        );
+    }
+    private get membershipServerSideExpiryTimeout(): number {
+        return (
+            this.membershipServerSideExpiryTimeoutOverride ??
+            this.joinConfig?.membershipServerSideExpiryTimeout ??
+            8_000
+        );
+    }
+    private get membershipKeepAlivePeriod(): number {
+        return this.joinConfig?.membershipKeepAlivePeriod ?? 5_000;
+    }
+    private get maximumRateLimitRetryCount(): number {
+        return this.joinConfig?.maximumRateLimitRetryCount ?? 10;
+    }
+    private get maximumNetworkErrorRetryCount(): number {
+        return this.joinConfig?.maximumNetworkErrorRetryCount ?? 10;
+    }
+
+    // Scheduler:
+    private oldStatus?: Status;
+    private scheduler = new ActionScheduler((type): Promise<ActionUpdate> => {
+        if (this.oldStatus) {
+            // we put this at the beginngin of the actsion scheduler loop handle callback since it is a loop this
+            // is equivalent to running it at the end of the loop. (just after applying the status/action list changes)
+            logger.debug(`MembershipManager applied action changes. Status: ${this.oldStatus} -> ${this.status}`);
+        }
+        this.oldStatus = this.status;
+        logger.debug(`MembershipManager before processing action. status=${this.oldStatus}`);
+        return this.membershipLoopHandler(type);
+    });
+
+    // LOOP HANDLER:
+    private async membershipLoopHandler(type: MembershipActionType): Promise<ActionUpdate> {
+        this.oldStatus = this.status;
+        switch (type) {
+            case MembershipActionType.SendFirstDelayedEvent: {
+                // Before we start we check if we come from a state where we have a delay id.
+                if (!this.state.delayId) {
+                    return this.sendFirstDelayedLeaveEvent(type); // Normal case without any previous delayed id.
+                } else {
+                    // This can happen if someone else (or another client) removes our own membership event.
+                    // It will trigger `onRTCSessionMemberUpdate` queue `MembershipActionType.SendFirstDelayedEvent`.
+                    // We might still have our delyed event from the previous participation and dependent on the serve this might not
+                    // get automatically removed if the state changes. Hence It would remove our membership unexpectedly shortly after the rejoin.
+                    //
+                    // In this block we will try to cancel this delayed event before setting up a new one.
+
+                    return this.cancelKnownDelayIdBeforeSendFirstDelayedEvent(type, this.state.delayId);
```

Comment:
The are a few places where the `type` argument is passed, but it appears to be redundant as we only ever call it with a single value `MembershipActionType.SendFirstDelayedEvent`.

Can the type argument be removed?

```suggestion
                    return this.cancelKnownDelayIdBeforeSendFirstDelayedEvent(this.state.delayId);
```